### PR TITLE
feat: support search route by label

### LINKED ISSUE #861: feat: Route group feature
# why
In order to allow users to filter out routes with a specified group from a large number of routes, I propose to add the route groups feature.

# how
## front-end
In order to finish this feature. There are 3 steps need to do.
- Adding group options to the edit or create route page.
image

- Route group list page
image

- Create or Edit route group page
image

## back-end
Create or update API to match front-end requirements.
- Create route and update route API need to update.
- Route Group List API.
- Create and update route group API.

# APISIX
Need to discuss how to implement route group requirements.

Free to comment here.

cc @juzhiyuan @liuxiran @gxthrj @nic-chen @membphis

### PR DESCRIPTION
**Pull request by starsz**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance

- Related issues

fix #861

### New feature or improvement
- Describe the details and related test reports.

1. Support search route by label.

### CODE REVIEW

File: api/internal/utils/utils_test.go [hunk 1]
```diff
@@ -42,3 +42,33 @@ func TestSumIPs_with_nil(t *testing.T) {
 	total := sumIPs(nil)
 	assert.Equal(t, uint16(0), total)
 }
+
+func TestGenLabelMap(t *testing.T) {
+	mp := GenLabelMap("l1")
```
Comment (nic-chen) [inline api/internal/utils/utils_test.go:47]: need test cases for illegal parameters, such as `,` ,  `,l2:,`

File: api/internal/handler/route/route_test.go [hunk 2]
```diff
@@ -797,6 +801,39 @@ func TestRoute(t *testing.T) {
...
+	err = json.Unmarshal([]byte(reqBody), listInput6)
+	assert.Nil(t, err)
+	ctx.SetInput(listInput6)
+	retPage, err = handler.List(ctx)
+	assert.Nil(t, err)
+	dataPage = retPage.(*store.ListOutput)
+	assert.Equal(t, len(dataPage.Rows), 1)
+
+	//list search label match
+	listInput7 := &ListInput{}
```
Comment (tokers) [inline api/internal/handler/route/route_test.go:827]: Should add another case that label specifies key and value, but value doesn't not matched with existing one.

File: api/internal/utils/utils.go [hunk 3]
```diff
@@ -94,3 +95,38 @@ func InterfaceToString(val interface{}) string {
...
+	}
+
+	labels := strings.Split(label, ",")
+	for _, l := range labels {
+		kv := strings.Split(l, ":")
+		if len(kv) == 2 {
+			mp[kv[0]] = kv[1]
+		} else if len(kv) == 1 {
+			mp[kv[0]] = ""
+		}
```
Comment (tokers) [inline api/internal/utils/utils.go:113]: Log the malformed label, it's useful when you debug problems.

Comment (starsz) [reply api/internal/utils/utils.go:113]: No, I think it's not a malformed label.It's a way only filter labels by key.

Comment (tokers) [reply api/internal/utils/utils.go:121]: I mean the `else` case.

Comment (starsz) [reply api/internal/utils/utils.go:125]: Fixed.

File: api/internal/utils/utils_test.go (hunk 1 above)
Comment (starsz) [reply api/internal/utils/utils_test.go]: Fixed.

**Review by tokers — APPROVED** [review]

**Review by LiteSun — APPROVED** [review]

**Pull request merged**